cranelift/codegen/src/opts/algebraic.isle

```diff
@@ -0,0 +1,189 @@
+;; Algebraic optimizations.
+
+;; uextend/sextend of a constant.
+(rule (simplify (uextend $I64 (iconst $I32 imm)))
+      (iconst $I64 imm))
+(rule (simplify (sextend $I64 (iconst $I32 (u64_from_imm64 imm))))
+      (iconst $I64 (imm64 (u64_sextend_u32 imm))))
+
+;; x+0 == 0+x == x.
+(rule 5 (simplify (iadd ty
+                      x
+                      (iconst ty (u64_from_imm64 0))))
+      (subsume x))
+(rule 4 (simplify (iadd ty
+                      (iconst ty (u64_from_imm64 0))
+                      x))
+      (subsume x))
+;; x-0 == x.
+(rule 8 (simplify (isub ty
+                      x
+                      (iconst ty (u64_from_imm64 0))))
+      (subsume x))
+;; 0-x == (ineg x).
+(rule 7 (simplify (isub ty
+                      (iconst ty (u64_from_imm64 0))
+                      x))
+      (ineg ty x))
+
+;; x*1 == 1*x == x.
+(rule 2 (simplify (imul ty
+                      x
+                      (iconst ty (u64_from_imm64 1))))
+      (subsume x))
+(rule 1 (simplify (imul ty
+                      (iconst ty (u64_from_imm64 1))
+                      x))
+      (subsume x))
+
+;; x/1 == x.
+(rule (simplify (sdiv ty
+                      x
+                      (iconst ty (u64_from_imm64 1))))
+      (subsume x))
+(rule (simplify (udiv ty
+                      x
+                      (iconst ty (u64_from_imm64 1))))
+      (subsume x))
+
+;; x>>0 == x<<0 == x rotr 0 == x rotl 0 == x.
+(rule 1 (simplify (ishl ty
+                      x
+                      (iconst ty (u64_from_imm64 0))))
+      (subsume x))
+(rule 1 (simplify (ushr ty
+                      x
+                      (iconst ty (u64_from_imm64 0))))
+      (subsume x))
+(rule 1 (simplify (sshr ty
+                      x
+                      (iconst ty (u64_from_imm64 0))))
+      (subsume x))
+(rule 1 (simplify (rotr ty
+                      x
+                      (iconst ty (u64_from_imm64 0))))
+      (subsume x))
+(rule 1 (simplify (rotl ty
+                      x
+                      (iconst ty (u64_from_imm64 0))))
+      (subsume x))
+
+;; x | 0 == 0 | x == x | x == x.
+(rule 2 (simplify (bor ty
+                     x
+                     (iconst ty (u64_from_imm64 0))))
+      (subsume x))
+(rule 1 (simplify (bor ty
+                     (iconst ty (u64_from_imm64 0))
+                     x))
+      (subsume x))
+(rule 0 (simplify (bor ty x x))
+      (subsume x))
+
+;; x ^ 0 == 0 ^ x == x.
+(rule 4 (simplify (bxor ty
+                     x
+                     (iconst ty (u64_from_imm64 0))))
+      (subsume x))
+(rule 3 (simplify (bxor ty
+                     (iconst ty (u64_from_imm64 0))
+                     x))
+      (subsume x))
+(rule 2 (simplify (bxor ty x x))
+      (subsume x))
```

Comment:
Good catch -- `x^x` does not actually equal `x` at all, except in one rare case :-)

> Multiplication by zero

Added!

> bitwise-and by zero

Added!

Reply:
Wait, so it _is_ okay to delete instructions that might trap? Where would I find something describing whether Cranelift is allowed to do optimizations like that?

If so you could add `x|not(x)=-1` and `x&not(x)=0`, I guess.

Reply:
Oh, no, we definitely don't delete any side-effecting instructions (aside from redundant loads but that's separate). I'm not sureI understand why including `x*0 == 0` or `x&0 == 0` has any effect on that?

Reply:
Okay, it's clear that these optimizations can potentially delete the only transitive data dependency on a potentially side-effecting instruction like integer division, right?

I guess the reason you're saying we don't delete those in spite of these optimizations is because they're kept even if there is no data dependency on them from any return instruction. Is that right?

In that case I think I'm satisfied. Though maybe a comment to that effect, in this source file specifically, would help somebody later.

Reply:
Right, side-effecting ops are kept in the top-level "side-effect skeleton", so they are never deleted.

Do you think we need a comment specifically on these rules, or ...? The reason I ask is that strictly speaking, *any optimization at all* that removes a reference to an input could trigger this confusion. I guess a top-level note on what rules are allowed to do (rewrite pure ops arbitrarily, remove references...) could help though.

Reply:
(Added a comment; feel free to suggest better clarifications if you have more ideas)